### DISCUSSION TITLE
How to find all actions/events that are able to create or modify a specific resource?

### DISCUSSION THREAD

**Discussion by michuer** [aws/aws-cli#8045] [Q&A]
Hi I'm looking for the way how to find all actions that are able to create or modify a specific resource.

Below is example, however, the question is more general and applies to **any resource**:
I tried to capture all events that are able to create/update api gateway stage for automatic WAF hooking. Initially, I identified 2 actions "CreateStage" and "UpdateStage" however, it turned out that apart from these 2 actions, the "CreateDeployment" action also is able to create stage without invoking the "CreateStage" event.

Unfortunately, logic implementations for AWS CLI functions are not publicly available so it's hard to find such information. 
In the CloudTrail history, you can search for all resources that have been affected by an event, but this is only a picture of the actions that have already been performed on the account, and I would like to predict which events may affect a particular resource.

**Comment by aBurmeseDev** [comment]
Hi @michuer - thanks for reaching out.

If I understand correctly, you're looking for a way to look up events with a given resource. If that's the case, I would suggest looking at AWS CLI [`lookup-events`](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/cloudtrail/lookup-events.html).

Here's more on how to use the method with `lookup-attributes` flag: https://repost.aws/knowledge-center/cloudtrail-search-actions

Hope that helps!

**Comment by uldyssian-sh** [comment]
You cannot reliably get a complete list of all actions that can create or modify a specific AWS resource using AWS CLI alone.

This is because AWS distinguishes between:
- what actually happened in your account, and
- what actions are theoretically allowed by permissions.

The practical approach combines CloudTrail and IAM analysis.

To find what actually created or modified the resource, use CloudTrail.

If the resource name is available in CloudTrail events:

aws cloudtrail lookup-events \
  --lookup-attributes AttributeKey=ResourceName,AttributeValue=<RESOURCE_NAME> \
  --max-results 50

If the resource name is not populated, filter by service instead:

aws cloudtrail lookup-events \
  --lookup-attributes AttributeKey=EventSource,AttributeValue=<service>.amazonaws.com \
  --max-results 50

To inspect the exact API calls and the calling principal, extract the raw event payload:

aws cloudtrail lookup-events \
  --lookup-attributes AttributeKey=EventSource,AttributeValue=<service>.amazonaws.com \
  --query "Events[].CloudTrailEvent" \
  --output text

To understand what actions could create or modify the resource, use the service IAM authorization reference.

AWS CLI cannot generically list “modify” actions, but IAM action names clearly indicate intent (Create, Put, Update, Modify, Delete). These actions define what is capable of changing a resource.

To determine who in your account can perform those actions, analyze effective permissions using IAM Access Analyzer or the IAM policy simulator.

Effective permissions are evaluated across:
- IAM identity policies
- Service Control Policies (SCPs)
- Permission boundaries
- Resource-based policies

In summary:
- CloudTrail tells you what actually modified the resource
- IAM documentation and analyzers tell you what actions could modify it and who is allowed

**Reply by uldyssian-sh** [reply]
Thanks! If this helps, please consider marking it as “Helpful” or “Accepted” 😊
